File: data-structures/sparse_table.py

from math import log2
from typing import Callable

class SparseTable:
    """
    - Consegue responder range minimum queries em O(1) [P.s: range minimum queries --> menor número dentro de um intervalo]
    - Não pode ser atualizada
    - construida em O(n log n)

    - como funciona: The main idea behind Sparse Tables is to precompute all answers for range queries with power of two length.
                     Afterwards a different range query can be answered by splitting the range into ranges with power of two lengths, 
                     looking up the precomputed answers, and combining them to receive a complete answer.
    """
    def __init__(self, arr: list, func,max_n: int):
        """
        func: função que depende do tipo de query.
        max_n: maior quantidade possível de elementos
        """

        k = int( log2(max_n) ) + 1

        self.st = [[0 for i in range(k)] for j in range(max_n)]

        self.func = func

        for i, v in enumerate(arr):
            self.st[i][0] = v

        for j in range(1, k):
            for i in range(len(arr)):
                if i + (1 << j) - 1 >= len(arr):
                    break

                self.st[i][j] = func( 
                    self.st[i][j-1], 
                    self.st[i + (1 << (j-1))][j-1] 
                )


    def query(self, L: int, R: int):
        # Assumes 1-indexed range

        R-=1; L-=1
        k = int(log2(max(R-L, 1)))

        return self.func( self.st[L][k], self.st[R - (1 << k) + 1][k] )

File: data-structures/sparse_table.cpp
// preprocessing: O(n log n)
// range minimum query  (minimum element in [L, R] interval): O(1)
struct SparseTable {
    vector<vector<int>> st;
    SparseTable(vector<int> &a) {
        int n = a.size();
        int k = log2(n);
        if ((1ll << k) < n) k++;
        st.assign(k+1, vector<int>(n, 0));
        copy(all(a), st[0].begin());

        for (int i = 1; i <= k; i++)
            for (int j = 0; j + (1 << i) <= n; j++)
                st[i][j] = min(st[i - 1][j], st[i - 1][j + (1 << (i - 1))]);
    }

    int query(int l, int r) {
        int i = log2(r - l + 1);
        return min(st[i][l], st[i][r - (1 << i) + 1]);
    }
};
